src/Functions/FunctionsHashing.h

```diff
@@ -1026,16 +1026,49 @@ class FunctionAnyHash : public IFunction
                 if constexpr (Impl::use_int_hash_for_pods)
                 {
                     if constexpr (std::is_same_v<ToType, UInt64>)
-                        h = IntHash64Impl::apply(bit_cast<UInt64>(vec_from[i]));
+                    {
+                        UInt64 v = bit_cast<UInt64>(vec_from[i]);
+                        if constexpr (std::endian::native == std::endian::big)
+                            v = __builtin_bswap64(v);
+                        h = IntHash64Impl::apply(v);
+                    }
                     else
-                        h = IntHash32Impl::apply(bit_cast<UInt32>(vec_from[i]));
+                    {
+                        UInt32 v = bit_cast<UInt32>(vec_from[i]);
+                        if constexpr (std::endian::native == std::endian::big)
+                            v = __builtin_bswap32(v);
+                        h = IntHash32Impl::apply(v);
+                    }
                 }
                 else
                 {
-                    if (std::is_same_v<Impl, JavaHashImpl>)
-                        h = JavaHashImpl::apply(vec_from[i]);
+                    if constexpr (std::endian::native == std::endian::little)
+                    {
+                        if (std::is_same_v<Impl, JavaHashImpl>)
+                            h = JavaHashImpl::apply(vec_from[i]);
+                        else
+                            h = apply(key, reinterpret_cast<const char *>(&vec_from[i]), sizeof(vec_from[i]));
+                    }
                     else
-                        h = apply(key, reinterpret_cast<const char *>(&vec_from[i]), sizeof(vec_from[i]));
+                    {
+                        if (std::is_same_v<Impl, JavaHashImpl>)
+                            h = JavaHashImpl::apply(vec_from[i]);
+                        else
+                        {
+                            if constexpr (std::is_same_v<ToType, UInt64>)
+                            {
+                                UInt64 v = bit_cast<UInt64>(vec_from[i]);
+                                v = __builtin_bswap64(v);
+                                h = apply(key, reinterpret_cast<const char *>(&v), sizeof(vec_from[i]));
+                            }
+                            else
+                            {
+                                UInt32 v = bit_cast<UInt32>(vec_from[i]);
+                                v = __builtin_bswap32(v);
+                                h = apply(key, reinterpret_cast<const char *>(&v), sizeof(vec_from[i]));
+                            }
+                        }
+                    }
```

Comment:
Sorry for yet another refactoring comment but it would be nice if the LE/BE distinction could generally be nested as deeply as possible into the if-else cascade. The control flow at high-level should be ideally driven by the hash functions and the data and not architecture details (endianness).

Basically the following. I added comments // (x) + explanation below.

```cpp
if constexpr (std::is_same_v<Impl, JavaHashImpl>) // (1)
    v = JavaHashImpl::apply(vec_from[i]); // (3)
else
{
    if constexpr (std::endian::native == std::endian::little)
        h = apply(key, reinterpret_cast<const char *>(&vec_from[i]), sizeof(vec_from[i]));
    else
    {
        if constexpr (std::is_same_v<ToType, UInt64>)
        {
            UInt64 v = bit_cast<UInt64>(vec_from[i]);
            v = __builtin_bswap64(v); // (4)
            h = apply(key, reinterpret_cast<const char *>(&v), sizeof(vec_from[i])); // (2)
        }
        else
        {
            UInt32 v = bit_cast<UInt32>(vec_from[i]);
            v = __builtin_bswap32(v);
            h = apply(key, reinterpret_cast<const char *>(&v), sizeof(vec_from[i]));
        }
    }
}
```

(1) would be cool to make that if `constexpr` (it isn't in the original code)
(2) for visual consistency, `sizeof(v)` instead of `sizeof(vec_from[i])` would be nice.
(3) just wondering why LE vs. BE makes no difference for JavaHash. Is that maybe an oversight?
(4) I think with C++23 std::byteswap, one would not need separate calls to `__builtin_bswap64` and `__builtin_bswap32` respectively - std::byteswap is templated on the input integer. Maybe add a TODO comment to use std::byteswap once that's available in ClickHouse?

Reply:
(1) Done.
(2) We cannot use sizeof(v) because vec_from[i] can be shorter than sizeof(v), for example it can be UInt16.
(3) I didn't touch JavaHash because it passes these functional tests on s390x: 00800_function_java_hash and00800_function_java_hash_with_unsigined_types.
(4) TODOs are not pretty in the code(just kidding), so let's leave it as is since builtin_bswap are also used in other places already.